충남대 FE 정지원 - 1단계 로그인 기능 구현하기

jjw5655 opened the pull request:
안녕하세요 우정민리뷰어님! 정지원이라고합니다. 잘 부탁드립니다.

구현 상세내용은 커밋메시지와 README.md파일 참고 부탁드립니다.

제가 리액트와 코딩실력이 많이 부족하여 상당부분 ai의 도움을 받아 미션을 진행하고 있습니다. ai가 짜준 코드의 전체적인 구조나 방식들 정도만 이해하고, "너가 구현한 코드에서 중요한 부분들 알려줘"라고 질문해서 주요 부분들을 더 살펴보는 식으로 공부하고 있습니다.
그래서 어떤 부분이 부족한지 뿐만 아니라, 어떤 부분이 잘 구현되었는지에 대해서도 모르는 부분이 많습니다. 이 부분들 많이많이 리뷰 남겨주시면 많은 도움이 될 것 같습니다. 감사합니다!

jjw5655 committed:
feat: 이메일 검증 유틸과 에러 컴포넌트 추가

- 이메일 형식 검증 함수 (isValidEmail, getEmailErrorMessage)
- 입력 검증 에러 메시지 컴포넌트 (ErrorMessage)
- utils/validation.ts 및 관련 export 정리

jjw5655 committed:
feat: ID 입력 필드에 이메일 검증 로직 적용

- LoginPage에 ID 검증 상태 관리 추가
- onBlur 시 검증 실행 및 에러 메시지 표시
- 에러 상태 시 Input 빨간 테두리 스타일 적용
- ErrorMessage 컴포넌트 연동

jjw5655 committed:
feat: 비밀번호 검증 유틸 추가

- 비밀번호 최소 8글자 검증 함수 (isValidPassword, getPasswordErrorMessage)
- utils/validation.ts 및 관련 export 정리

jjw5655 committed:
feat: PW 입력 필드에 비밀번호 검증 로직 적용

- LoginPage에 PW 검증 상태 관리 추가
- onBlur 시 검증 실행 및 에러 메시지 표시
- 에러 상태 시 Input 빨간 테두리 스타일 적용
- ErrorMessage 컴포넌트 연동

jjw5655 committed:
feat: 로그인 버튼 활성화 조건 구현

- ID와 PW 검증 상태를 종합하여 버튼 활성화/비활성화
- 실시간 유효성 검증으로 즉시 버튼 상태 반영
- 모든 조건 충족 시에만 로그인 가능

jjw5655 committed:
refactor: useLoginForm 커스텀 훅으로 로직 분리

- 모든 검증 로직과 상태를 useLoginForm으로 이동
- LoginPage에서 useState 완전 제거
- 재사용 가능한 로그인 폼 로직 구성
- UI와 비즈니스 로직 완전 분리

jjw5655 committed:
style: .prettierignore 파일 끝에 EOL 추가

jjw5655 committed:
docs: README.md에 미션 2-1단계 구현사항 요약 추가

greenblues1190 commented on src/utils/validation.ts:
이메일과 패스워드에 대한 정규표현식이나 검증처리가 컴포넌트 파일에 함께 존재하면 코드 가독성이 많이 저해되는데, 잘 추상화해주셨어요!

greenblues1190 commented on src/hooks/useLoginForm.ts:
```diff
@@ -0,0 +1,74 @@
+  const handlePasswordBlur = () => {
+    const errorMessage = getPasswordErrorMessage(password);
+    setPasswordError(errorMessage);
+  };
```
**[변경 요청]** 한번 blur 이벤트가 일어난 입력 필드에 대해서는 change 이벤트가 일어날 때도 검증을 하면 다음 요구사항을 충족할 수 있을 것 같아요.

> 위 조건이 충족되면 에러 메시지가 사라져요.

`isEmailTouched`, `isPasswordTouched`라는 리액트 상태를 만들어서, blur가 한번이라도 발생한 입력 필드는 change에서도 검증해서 사용자가 바로 피드백을 받도록 해보시면 어떨까요?

jjw5655 replied:
한번 blur가 발생한 입력 필드는 이후 타이핑할 때마다 실시간으로 검증하여 에러 메시지가 즉시 사라지도록 개선했습니다.

greenblues1190 commented on src/pages/LoginPage.tsx:
**[참고]** 폼 제출 로직 구현 시 브라우저 내장 로직을 활용하여 간결하게 작성할 수도 있어요.

- [\<form\>](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement)
    - onSubmit
    - [FormData](https://developer.mozilla.org/ko/docs/Web/API/FormData)
-  [Constraint Validation API](https://developer.mozilla.org/en-US/docs/Web/HTML/Guides/Constraint_validation)
    - checkValidity()
    - reportValidity()

greenblues1190 commented on src/hooks/useLoginForm.ts:
```diff
@@ -0,0 +1,74 @@
+
+    const from = location.state?.from || '/';
+    navigate(from, { replace: true });
+  };
```
**[변경 요청]** 지금 로그인 **폼** 훅 안에서 navigate()로 화면 이동까지 하고 계신데요. 폼 제출 후 로직들이 폼 훅안에서 처리할 필요가 있을지에 대해 고민해보시면 좋겠어요. 하나의 훅, 함수는 하나의 기능과 역할만을 가지는 것이 수정, 확장에 유리합니다.

`useLoginForm` 훅은 이메일, 비밀번호 값 관리와 유효성 검증에만 집중하고,
네비게이션과 같은 로그인 폼 제출 후 로직은 `LoginPage`에서 처리해보시면 어떨까요?

예를 들어

- useLoginForm에서 onSubmit 반환.
    - onSubmit은 콜백 함수를 받아, 폼에 입력한 id, 패스워드가 유효할 경우 폼 데이터를 인자로 자동으로 전달하고 호출하는 역할
    ```
    const { onSubmit } = useLoginForm();
    ```
    
- LoginPage에서 반환된 값을 받아 navigate() 실행
    ```
    <form submit={onSubmit(({ id, password }) => { /* 로그인 처리*/ })}>...</form>
    ```

jjw5655 replied:
useLoginForm 훅에서 화면 이동 로직을 분리하여 폼 검증만 담당하고, LoginPage에서 로그인 성공 후 처리를 담당하도록 수정했습니다. 피드백 감사합니다 👍

greenblues1190 reviewed (changes requested):
안녕하세요 지원님!
전체적으로 훅, 유틸, 컴포넌트를 깔끔히 분리해주셔서 구조가 명확하고 재사용성이 높아졌어요.
validation 로직과 ErrorMessage 컴포넌트로 추상화한 부분도 좋았습니다 👍

머지 전 [변경 요청]으로 남긴 사항들 반영 부탁드릴게요!

jjw5655 committed:
feat: 로그인 폼 점진적 검증 기능 추가

- blur 발생 후 change 이벤트에서도 실시간 검증 수행
- 사용자가 에러를 수정하는 즉시 피드백 제공하여 UX 개선

jjw5655 committed:
refactor: useLoginForm 훅에서 네비게이션 로직 분리

- 로그인 성공 후 처리를 LoginPage 컴포넌트로 이동
- onSubmit 콜백 패턴으로 재사용성 개선
- form 태그와 submit 버튼으로 표준 폼 구조 적용

greenblues1190 commented on src/hooks/useLoginForm.ts:
```diff
@@ -0,0 +1,94 @@
+    !emailError &&
+    !passwordError;
+
+  const onSubmit =
+    (callback: (data: LoginFormData) => void) => (e: React.FormEvent) => {
+      e.preventDefault();
+
+      if (!isFormValid) {
+        return;
+      }
+
+      callback({ email, password });
+    };
```
onSubmit에서 callback을 주입받을 수 있게하여 폼의 코어 로직을 추상화해주신 점이 좋네요 👍

greenblues1190 commented on src/hooks/useLoginForm.ts:
```diff
@@ -0,0 +1,94 @@
+}
+
+export const useLoginForm = (): UseLoginFormReturn => {
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+  const [emailError, setEmailError] = useState<string | null>(null);
+  const [passwordError, setPasswordError] = useState<string | null>(null);
+  const [isEmailTouched, setIsEmailTouched] = useState(false);
+  const [isPasswordTouched, setIsPasswordTouched] = useState(false);
+
+  const handleEmailChange = (value: string) => {
+    setEmail(value);
+    if (isEmailTouched) {
+      const errorMessage = getEmailErrorMessage(value);
+      setEmailError(errorMessage);
+    }
+  };
+
+  const handlePasswordChange = (value: string) => {
+    setPassword(value);
+    if (isPasswordTouched) {
+      const errorMessage = getPasswordErrorMessage(value);
+      setPasswordError(errorMessage);
+    }
+  };
+
+  const handleEmailBlur = () => {
+    setIsEmailTouched(true);
+    const errorMessage = getEmailErrorMessage(email);
+    setEmailError(errorMessage);
+  };
+
+  const handlePasswordBlur = () => {
+    setIsPasswordTouched(true);
+    const errorMessage = getPasswordErrorMessage(password);
+    setPasswordError(errorMessage);
+  };
```
서비스 전반적으로 useForm이라는 공통 폼 관리 훅을 재사용해서 써야한다면, useLoginForm을 어떻게 수정해야할지도 고민해보셨으면 좋겠습니다!

greenblues1190 reviewed (approved):
안녕하세요 지원님! 고생하셨습니다. 로그인 기능이 잘 동작하는 것도 확인했습니다~
추가로 이런 것을 고민해보셨으면 좋겠다 하는 것들을 코멘트로 남겼어요!
2단계도 화이팅입니다 😄
